<a href="https://colab.research.google.com/github/ncbpsicologia/Bootcamp_ITacademy_DataAnalyst/blob/main/Sprint_7_Natalia_Cuenca.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **NIVELL X**
## _EXERCICI X, TÍTOL_

In [ ]:
# PASO n
# EXPLICACIÓN DE LA FUNCIÓN
# CÓDIGO DE LA FUNCIÓN

In [ ]:
# COMPROBACIÓN DEL PASO n

# **NIVELL 1**
## _EXERCICI 1, SELECCIÓ DE PERSONAL_

Tu treballes com a analista de dades en el departament de RRHH d'una empresa. S'ha obert una nova oferta laboral interna i l'empresa ha decidit que només les primeres 10 persones que s'inscriguin podran participar en el procés de selecció.
Per això, t'han encarregat desenvolupar una eina senzilla en Python que actuï com a formulari d'inscripció i sistema de classificació inicial.
>
Aquest formulari ha de permetre introduir, per a cada candidat/a:
- Nom i Cognoms
- Telèfon de contacte
- Anys d'experiència dins de l'empresa
- Coneixement de Català (0–3 punts)
0 = sense coneixement
1 = fins a nivell B1
2 = nivell B2
3 = C1 o superior
- Domini d'Excel (0–3 punts)
0 = nivell bàsic
1 = ús de funcions
2 = taules dinàmiques
3 = Visual Basic i Macros
- Títol acadèmic (0–3 punts)
0 = ESO
1 = FP
2 = Universitat
3 = Màster o superior

>>
Assegurant-se que tota la informació introduïda és correcta i entra dins els valors permesos.
A partir d'aquestes dades, el programa haurà de calcular una puntuació total aplicant el pes de l'experiència (anys * 1.5) i sumant-hi els punts de català, excel i titulació, i a continuació classificar automàticament cada persona segons els llindars establerts per l'empresa:
- Prioritat Alta → puntuació ≥ 12
- Candidat adequat → 8 ≤ puntuació < 12
- No prioritari → puntuació < 8

Cada vegada que algú complet a el formulari, s'ha de mostrar la seva puntuació individual. Quan ja s'hagin inscrit exactament deu persones, l'eina ha de deixar d'acceptar nous registres i generar un informe final en pantalla, ordenat primer segons la prioritat i, en cas d'empat, per la puntuació obtinguda. Aquest informe ha d'incloure el nom, el telèfon i la puntuació de cada candidat per tal que el departament de RRHH els pugui contactar ràpidament.

> Important
Per resoldre aquest exercici, hauràs de:
** Desenvolupar un sistema modular amb les funcions següents: **
- obtenir_dades
- validar_dades
- calcular_puntuacio
- classificar_candidat
- mostrar_individual
- mostrar_informe
- main_seleccio # orquestra totes les funcions

** - Escriure primer l'algoritme pas a pas, indicant clarament quines operacions farà cada funció. **
** - Escriure el codi de cada funció. **

>> Per fer proves:

inputs_test = [
    "Laura Serra", "608112233", 6, 3, 3, 3,
    "Clara Gómez", "655221199", 3, 3, 3, 2,
    "Sergio López", "644112211", 3, 2, 2, 2,
    "Candidat Error", "611000000", 5, 7, 3, 2, # Català = 7 → NO VÀLID
    "Marta Ruiz", "699887766", 2, 3, 2, 1,
    "Marc Vidal", "612443322", 6, 3, 3, 3,
    "Júlia Romero", "677998877", 5, 3, 3, 2,
    "Pol Navarro", "699554433", 4, 3, 2, 2,
    "David Ferrer", "677665544", 2, 1, 1, 1,
    "Anna Paredes", "633221144", 1, 2, 1, 0,  
    "Anibal Paredes", "644272727", 1, 2, 1, 0,  
    "Núria Casals", "600112200", 8, 2, 3, 3 # 11è candidat     
]

In [ ]:
# PASO 1: OBTENER DATOS

# El objetivo de esta función es obtener datos básicos sobre la persona usuaria.
# La función recibirá los datos a través de inputs (un input por dato), los almacenará en variables y retornará los datos introducidos en un diccionario.
# Concretamente, se pide: nombre y apellidos, teléfono de contacto, años de experiencia en la empresa, nivel de catalán (entre 0 y 3), nivel de Excel (entre 0 y 3) y nivel de titulación académica (entre 0 y 3)

def obtenir_dades():
    nom = input("Escriu el teu nom i cognoms: ")
    tlf = input("Telèfon de contacte: ")
    exp = int(input("Introdueix quants anys d'experiència tens dins l'empresa: "))
    cat = int(input(
        "Indica el teu nivell de català (0-3)\n"
        "0 = sense coneixement\n"
        "1 = fins a nivell B1\n"
        "2 = nivell B2\n"
        "3 = nivell C1 o superior\n"
        "Valor: "
        ))
    exl = int(input(
        "Indica el teu nivell de domini d'Excel (0-3)\n"
        "0 = nivell bàsic\n"
        "1 = ús de funcions\n"
        "2 = taules dinàmiques\n"
        "3 = Visual Basic i Macros\n"
        "Valor: "
        ))
    tit = int(input(
        "Indica el teu nivell acadèmic (0-3)\n"
        "0 = ESO\n"
        "1 = FP\n"
        "2 = Universitat\n"
        "3 = Màster o superior\n"
        "Valor: "
       ))

    candidatura = {
        "nom": nom,
        "telefon": tlf,
        "experiencia": exp,
        "catala": cat,
        "excel": exl,
        "titulacio": tit
        }

    return candidatura

In [ ]:
# COMPROBACIÓN DEL PASO 1

obtenir_dades()

Escriu el teu nom i cognoms: Laura Serra
Telèfon de contacte: 608112233
Introdueix quants anys d'experiència tens dins l'empresa: 6
Indica el teu nivell de català (0-3)
0 = sense coneixement
1 = fins a nivell B1
2 = nivell B2
3 = nivell C1 o superior
Valor: 3
Indica el teu nivell de domini d'Excel (0-3)
0 = nivell bàsic
1 = ús de funcions
2 = taules dinàmiques
3 = Visual Basic i Macros
Valor: 3
Indica el teu nivell acadèmic (0-3)
0 = ESO
1 = FP
2 = Universitat
3 = Màster o superior
Valor: 3


{'nom': 'Laura Serra',
 'telefon': '608112233',
 'experiencia': 6,
 'catala': 3,
 'excel': 3,
 'titulacio': 3}

In [ ]:
# PASO 2: VALIDAR DATOS

# El objetivo de esta función es validar los datos obtenidos, comprobando que sean correctos/adecuados. Recibe el diccionario que hemos creado.
# La función comprobará distintas condiciones a través de condicionales. Si algun dato es incorrecto, rechazará el registro. Si todos los datos son correctos, retornará que la validación es correcta.
# Concretamente, se comprobará que los años de experiencia no sean negativos, y que el nivel de catalan, de Excel y de titulación académica tengan un valor entre 0 y 3

def validar_dades(candidatura):
    if candidatura["experiencia"] < 0:
        print("Error: els anys d'experiència no poden ser negatius.")
        return False

    if candidatura["catala"] < 0 or candidatura["catala"] > 3:
        print("Error: el nivell de català ha de ser un valor entre 0 i 3.")
        return False

    if candidatura["excel"] < 0 or candidatura["excel"] > 3:
        print("Error: el nivell d'Excel ha de ser un valor entre 0 i 3.")
        return False

    if candidatura["titulacio"] < 0 or candidatura["titulacio"] > 3:
        print("Error: el nivell de titulació ha de ser un valor entre 0 i 3.")
        return False

    print("\nRegistre acceptat.")
    return True

In [ ]:
# COMPROBACIÓN 1 DEL PASO 2

candidatura = obtenir_dades()

validar_dades(candidatura)

Escriu el teu nom i cognoms: Laura Serra
Telèfon de contacte: 608112233
Introdueix quants anys d'experiència tens dins l'empresa: 6
Indica el teu nivell de català (0-3)
0 = sense coneixement
1 = fins a nivell B1
2 = nivell B2
3 = nivell C1 o superior
Valor: 3
Indica el teu nivell de domini d'Excel (0-3)
0 = nivell bàsic
1 = ús de funcions
2 = taules dinàmiques
3 = Visual Basic i Macros
Valor: 3
Indica el teu nivell acadèmic (0-3)
0 = ESO
1 = FP
2 = Universitat
3 = Màster o superior
Valor: 3

Registre acceptat.


True

In [ ]:
# COMPROBACIÓN 2 DEL PASO 2

candidatura = obtenir_dades()

validar_dades(candidatura)

Escriu el teu nom i cognoms: Candidat Error
Telèfon de contacte: 611000000
Introdueix quants anys d'experiència tens dins l'empresa: 5
Indica el teu nivell de català (0-3)
0 = sense coneixement
1 = fins a nivell B1
2 = nivell B2
3 = nivell C1 o superior
Valor: 7
Indica el teu nivell de domini d'Excel (0-3)
0 = nivell bàsic
1 = ús de funcions
2 = taules dinàmiques
3 = Visual Basic i Macros
Valor: 3
Indica el teu nivell acadèmic (0-3)
0 = ESO
1 = FP
2 = Universitat
3 = Màster o superior
Valor: 2
Error: el nivell de català ha de ser un valor entre 0 i 3.


False

In [ ]:
# PASO 3: CALCULAR PUNTUACIÓN

# El objetivo de esta función es calcular la puntuación de la persona usuaria. Recibe el diccionario que hemos creado.
# La función realizará los cálculos pertinentes a través de los operadores y retornará la puntuación total
# Concretamente, la puntuación se basará en los años de experiencia en la empresa (con un peso significativo), el nivel de catalán, de dominio de Excel y de titulación académica.

def calcular_puntuacio(candidatura):
  puntuacio = (
      candidatura["experiencia"] * 1.5
      + candidatura["catala"]
      + candidatura["excel"]
      + candidatura["titulacio"]
  )

  return puntuacio

In [ ]:
# COMPROBACIÓN DEL PASO 3

puntuacio = calcular_puntuacio(candidatura)
print("La puntuació és:", puntuacio)
candidatura["puntuacio"] = puntuacio

La puntuació és: 19.5


In [ ]:
# PASO 4: CLASIFICAR CANDIDATOS

# El objetivo de esta función es asignar una etiqueta a cada persona usuaria en base a su puntuacion, clasificarlas. Recibe la puntuacion calculada en el paso anterior.
# La función asignará y retornará la etiqueta correspondiente a través de condicionales.
# Concretamente, las etiquetas seran: Prioritat Alta → puntuació ≥ 12, Candidat adequat → 8 ≤ puntuació < 12, No prioritari → puntuació < 8

def clasificar_candidat(puntuacio):
  if puntuacio >= 12:
    return "Prioritat alta"

  elif puntuacio >= 8:
    return "Candidat/da adequat/da"

  else:
    return "No prioritat"

In [ ]:
# COMPROBACIÓN 1 DEL PASO 4

classificacio = clasificar_candidat(puntuacio)
print("L'etiqueta de classificació és:", classificacio)
candidatura["classificacio"] = classificacio

L'etiqueta de classificació és: Prioritat alta


In [ ]:
# PASO 5: MOSTRAR INFORMACIÓN INDIVIDUAL

# El objetivo de esta función es mostrar un resumen del resultado de la candidatura. Recibe el diccionario que hemos creado (para la variable nom) y las variables puntuacio y clasificacio creadas después.
# La función mostrará en pantalla concretamente los siguientes datos: nombre del/a candidato/a, puntuación obtenida y etiqueta de clasificación asignada.

def mostrar_individual(candidatura, puntuacio, classificacio):
    print("--RESULTAT DEL/A CANDIDAT/A--")
    print(f"Nom: {candidatura['nom']}")
    print(f"Puntuació: {puntuacio}")
    print(f"Classificació: {classificacio}")

In [ ]:
# COMPROBACIÓN DEL PASO 5

mostrar_individual(candidatura, puntuacio, classificacio)

--RESULTAT DEL/A CANDIDAT/A--
Nom: Candidat Error
Puntuació: 19.5
Classificació: Prioritat alta


In [ ]:
# PASO 6: MOSTRAR INFORME

# El objetivo de esta función es mostrar los candidatos por orden de prioridad. Recibe el listado de candidaturas (un diccionario por cada persona)
# La función recorrerá los registros con un bucle for, ordenará los candidatos y los mostrará en pantalla, concretamente los siguientes datos: nombre del/a candidato/a, teléfono, puntuación obtenida y etiqueta de clasificación asignada.

def mostrar_informe(listado_candidaturas):

    listado_candidaturas.sort(key=lambda c: c["puntuacio"], reverse=True)

    print("\n===== INFORME FINAL =====\n")

    for c in listado_candidaturas:
        print(f"Nom: {c['nom']}")
        print(f"Telèfon: {c['telefon']}")
        print(f"Puntuació: {c['puntuacio']}")
        print(f"Classificació: {c['classificacio']}")
        print("-------------------")


In [ ]:
# COMPROBACIÓN DEL PASO 6

listado_candidaturas_test = [
    {
        "nom": "Laura Serra",
        "telefon": "608112233",
        "puntuacio": 18,
        "classificacio": "Prioritat alta"
    },
    {
        "nom": "Clara Gómez",
        "telefon": "655221199",
        "puntuacio": 12,
        "classificacio": "Candidat/da adequat/da"
    },
    {
        "nom": "Sergio López",
        "telefon": "644112211",
        "puntuacio": 10,
        "classificacio": "Candidat/da adequat/da"
    },
    {
        "nom": "Candidat Error",
        "telefon": "611000000",
        "puntuacio": 8,
        "classificacio": "No prioritari"
    }
]

mostrar_informe(listado_candidaturas_test)


===== INFORME FINAL =====

Nom: Laura Serra
Telèfon: 608112233
Puntuació: 18
Classificació: Prioritat alta
-------------------
Nom: Clara Gómez
Telèfon: 655221199
Puntuació: 12
Classificació: Candidat/da adequat/da
-------------------
Nom: Sergio López
Telèfon: 644112211
Puntuació: 10
Classificació: Candidat/da adequat/da
-------------------
Nom: Candidat Error
Telèfon: 611000000
Puntuació: 8
Classificació: No prioritari
-------------------


In [ ]:
# PASO 7: FUNCIÓN MAIN (ORQUESTADORA)

# El objetivo de esta función es orquestar el resto de pasos. Como el código es largo, los voy a ir describiendo a lo largo de la función, por facilitar la legibilidad:

def main_seleccio():

    # Crear una lista vacía para guardar las candiadturas válidas
    llista_candidats = []

    # Repetir el proceso hasta obtener 10 candidaturas válidas
    while len(llista_candidats) < 10:

        # PASO 1: obtener datos
        candidatura = obtenir_dades()

        # PASO 2: validar datos
        if validar_dades(candidatura):

            # PASO 3: calcular puntuación
            puntuacio = calcular_puntuacio(candidatura)

            # PASO 4: clasificar
            classificacio = clasificar_candidat(puntuacio)

            # guardar resultados dentro del diccionario del/a candidato/a
            candidatura["puntuacio"] = puntuacio
            candidatura["classificacio"] = classificacio

            # añadir a la lista final
            llista_candidats.append(candidatura)

            # mostrar resultado individual
            mostrar_individual(candidatura, puntuacio, classificacio)

        else:
            print("Candidatura rebutjada. Torna-ho a intentar.\n")

    # PASO 6: informe final
    mostrar_informe(llista_candidats)

In [ ]:
# COMPROBACIÓN DEL PASO 7

main_seleccio()

Escriu el teu nom i cognoms: Laura Serra
Telèfon de contacte: 608112233
Introdueix quants anys d'experiència tens dins l'empresa: 6
Indica el teu nivell de català (0-3)
0 = sense coneixement
1 = fins a nivell B1
2 = nivell B2
3 = nivell C1 o superior
Valor: 3
Indica el teu nivell de domini d'Excel (0-3)
0 = nivell bàsic
1 = ús de funcions
2 = taules dinàmiques
3 = Visual Basic i Macros
Valor: 3
Indica el teu nivell acadèmic (0-3)
0 = ESO
1 = FP
2 = Universitat
3 = Màster o superior
Valor: 3

Registre acceptat.
--RESULTAT DEL/A CANDIDAT/A--
Nom: Laura Serra
Puntuació: 18.0
Classificació: Prioritat alta
Escriu el teu nom i cognoms: Clara Gómez
Telèfon de contacte: 655221199
Introdueix quants anys d'experiència tens dins l'empresa: 3
Indica el teu nivell de català (0-3)
0 = sense coneixement
1 = fins a nivell B1
2 = nivell B2
3 = nivell C1 o superior
Valor: 3
Indica el teu nivell de domini d'Excel (0-3)
0 = nivell bàsic
1 = ús de funcions
2 = taules dinàmiques
3 = Visual Basic i Macros
Va

# **NIVELL 1**
## _EXERCICI 2, SEGMENTACIÓ PER RISC DE CISTELLA DE COMPRA_

Treballes com a analista de dades al departament de Màrqueting Digital.
L'empresa vol millorar l'efectivitat de les campanyes de recuperació de cistellas de compra abandonats i necessita identificar quins clients tenen més risc d'abandonament, quin segment de comportament presenten i quin missatge publicitari és el més adequat per a cadascun.

**Disposes d'un conjunt de dades amb informació bàsica de 10 clients:**
- id_client
- nom
- edat
- grup_edat
- codi_comportament
              
> clients = [
    {"id_client": 1, "nom": "Laura Serra", "edat": 23, "grup_edat": "18-25", "codi_comportament": 4},
    {"id_client": 2, "nom": "Marc Vidal", "edat": 35, "grup_edat": "26-40", "codi_comportament": 3},
    {"id_client": 3, "nom": "Ana López", "edat": 29, "grup_edat": "26-40", "codi_comportament": 2},
    {"id_client": 4, "nom": "Joan Riera", "edat": 19, "grup_edat": "18-25", "codi_comportament": 2},
    {"id_client": 5, "nom": "Mónica Pérez", "edat": 47, "grup_edat": "41-60", "codi_comportament": 4},
    {"id_client": 6, "nom": "Luis García", "edat": 52, "grup_edat": "41-60", "codi_comportament": 1},
    {"id_client": 7, "nom": "Pilar Sánchez", "edat": 61, "grup_edat": "60+", "codi_comportament": 3},
    {"id_client": 8, "nom": "Eva Martín", "edat": 38, "grup_edat": "26-40", "codi_comportament": 0},
    {"id_client": 9, "nom": "Diego Romero", "edat": 24, "grup_edat": "18-25", "codi_comportament": 1},
    {"id_client": 10, "nom": "Núria Costa", "edat": 33, "grup_edat": "26-40", "codi_comportament": 4}
]
       
**El departament de CRM ha definit manualment:**
> Codis de comportament
- 0 → Navega però no afegeix productes
- 1 → Afegeix productes però no arriba al checkout
- 2 → Arriba al checkout però no paga
- 3 → Cistella de compra abandonat fa menys de 24 h
- 4 → Cistella de compra abandonat fa més de 24 h

> Factors segons codi de comportament
- Codi 0 → 0.10
- Codi 1 → 0.25
- Codi 2 → 0.50
- Codi 3 → 0.75
- Codi 4 → 0.90

> Factors segons grup d'edat
- 18–25 → 1.20
- 26–40 → 1.00
- 41–60 → 0.85
- 60+ → 0.70

**Fórmula del risc final**
risc_final = factor codi * factor edat

Un cop processats els clients, has de ser capaç de respondre:
» 1. Quins clients s'han de contactar primer en una campanya de retargeting?
» 2. Quines diferències hi ha entre clients amb el mateix comportament però d'edats diferents? (Per exemple: un client jove que abandona la cistella de compra fa 24 h té el mateix risc que un de més edat?)
» 3. Quins clients presenten tan poc interès que no val la pena invertir-hi pressupost?
» 4. Quin client seria el més rendible de recuperar?
Interpretant el risc final i el segment.

⚙️ Requisits tècnics del teu programa
Hauràs de construir un sistema modular utilitzant diverses funcions i, sobretot, sense utilitzar if/elif/else per fer la segmentació.

**Important**
Per resoldre aquest exercici, hauràs de:
Desenvolupar un sistema modular amb les funcions següents:
S'ha d'implementar amb un diccionari de diccionaris, sense condicions.

- processar_clients # s'ha d'alimentar de les funcions següents:
- obtenir_segment_i_factor
- obtenir_factor_edat
- calcular_risc
- mostrar_dataframe o mostrar_diccionari
- mostrar_respostes
- main_marketing # orquestra totes les funcions

» Escriure primer l'algoritme pas a pas, indicant clarament quines operacions farà cada funció.
» Escriure el codi de cada funció.

✅ Resultat esperat
Un DataFrame o diccionari final amb:
- id_cliente
- nombre
- edad
- grupo_edad
- codigo_comportamiento
- segment
- missatge
- factor_codi
- factor_edat
- risc_final

I capacitat per respondre totes les preguntes de màrqueting demanades en una funció.

In [ ]:
# PASO "0": INFORMACIÓN Y CREACIÓN DE DICCIONARIOS

clients = [
    {"id_client": 1, "nom": "Laura Serra", "edat": 23, "grup_edat": "18-25", "codi_comportament": 4},
    {"id_client": 2, "nom": "Marc Vidal", "edat": 35, "grup_edat": "26-40", "codi_comportament": 3},
    {"id_client": 3, "nom": "Ana López", "edat": 29, "grup_edat": "26-40", "codi_comportament": 2},
    {"id_client": 4, "nom": "Joan Riera", "edat": 19, "grup_edat": "18-25", "codi_comportament": 2},
    {"id_client": 5, "nom": "Mónica Pérez", "edat": 47, "grup_edat": "41-60", "codi_comportament": 4},
    {"id_client": 6, "nom": "Luis García", "edat": 52, "grup_edat": "41-60", "codi_comportament": 1},
    {"id_client": 7, "nom": "Pilar Sánchez", "edat": 61, "grup_edat": "60+", "codi_comportament": 3},
    {"id_client": 8, "nom": "Eva Martín", "edat": 38, "grup_edat": "26-40", "codi_comportament": 0},
    {"id_client": 9, "nom": "Diego Romero", "edat": 24, "grup_edat": "18-25", "codi_comportament": 1},
    {"id_client": 10, "nom": "Núria Costa", "edat": 33, "grup_edat": "26-40", "codi_comportament": 4}
            ]


codis_comportament = {
    0 : "Navega però no afegeix productes",
    1 : "Afegeix productes però no arriba al checkout",
    2 : "Arriba al checkout però no paga",
    3 : "Cistella de compra abandonat fa menys de 24 h",
    4 : "Cistella de compra abandonat fa més de 24 h"
                    }

factors_codi = {
    0 : 0.10,
    1 : 0.25,
    2 : 0.50,
    3 : 0.75,
    4 : 0.90
              }

factors_edat = {
    "18-25" : 1.20,
    "26-40" : 1.00,
    "41-60" : 0.85,
    "60+" : 0.70,
              }

In [ ]:
# PASO 1 OBTENER FACTOR EDAD

# El objetivo de la función es convertir el grupo de edad en un factor numérico. Recibe un grupo de edad.
# La función hará la conversión a través de los diccionarios. Retornará un valor numérico.

def obtenir_factor_edat(grup_edat):
  return factors_edat[grup_edat]

In [ ]:
# COMPROBACIONES DEL PASO 1

print(f"El factor_edat per la franja '18-25' és: {obtenir_factor_edat("18-25")}")
print(f"El factor_edat per la franja '26-40' és: {obtenir_factor_edat("26-40")}")
print(f"El factor_edat per la franja '41-60' és: {obtenir_factor_edat("41-60")}")
print(f"El factor_edat per la franja '60+' és: {obtenir_factor_edat("60+")}")

El factor_edat per la franja '18-25' és: 1.2
El factor_edat per la franja '26-40' és: 1.0
El factor_edat per la franja '41-60' és: 0.85
El factor_edat per la franja '60+' és: 0.7


In [ ]:
# PASO 2 OBTENER FACTOR Y CÓDIGO

# El objetivo de la función es convertir el código de comportamiento numérico en un segmento descriptivo y un factor de riesgo numérico. Recibe el código de comportamiento.
# La función hará la conversión a través de los diccionarios. Retornará un valor numérico y su descripción.

def obtenir_segment_factor(codi_comportament):
    return {
        0: {"segment": "Navega però no afegeix productes", "factor": 0.10},
        1: {"segment": "Afegeix productes però no checkout", "factor": 0.25},
        2: {"segment": "Arriba al checkout però no paga", "factor": 0.50},
        3: {"segment": "Cistella abandonada <24h", "factor": 0.75},
        4: {"segment": "Cistella abandonada >24h", "factor": 0.90}
    }[codi_comportament]

In [ ]:
# COMPROBACIONES DEL PASO 2

print(f"El codi de comportament per el valor '0' és: {obtenir_segment_factor(0)}")
print(f"El codi de comportament per el valor '1' és: {obtenir_segment_factor(1)}")
print(f"El codi de comportament per el valor '2' és: {obtenir_segment_factor(2)}")
print(f"El codi de comportament per el valor '3' és: {obtenir_segment_factor(3)}")
print(f"El codi de comportament per el valor '4' és: {obtenir_segment_factor(4)}")

El codi de comportament per el valor '0' és: {'segment': 'Navega però no afegeix productes', 'factor': 0.1}
El codi de comportament per el valor '1' és: {'segment': 'Afegeix productes però no checkout', 'factor': 0.25}
El codi de comportament per el valor '2' és: {'segment': 'Arriba al checkout però no paga', 'factor': 0.5}
El codi de comportament per el valor '3' és: {'segment': 'Cistella abandonada <24h', 'factor': 0.75}
El codi de comportament per el valor '4' és: {'segment': 'Cistella abandonada >24h', 'factor': 0.9}


In [ ]:
# PASO 3: CALCULAR EL RIESGO

# El objetivo de la función es calcular el riesgo final de cada cliente/a. Recibe dos datos numéricos: factor_codi y factor_edat
# La función realizará el cálculo siguiente: multiplicar el factor_codi por el factor_edat. Retornará un valor numérico.

def calcular_risc(factor_codi, factor_edat):
    return factor_codi * factor_edat

In [ ]:
# COMPROBACIONES DEL PASO 3

print(f"El risc d'aquest/a client/a és: {calcular_risc(0.90, 1.20)}")
print(f"El risc d'aquest/a client/a és: {calcular_risc(0.50, 1.00)}")

El risc d'aquest/a client/a és: 1.08
El risc d'aquest/a client/a és: 0.5


In [ ]:
# PASO 4: PROCESAR CLIENTES

# El objetivo de la función es combinar la información recabada hasta ahora. Recibe la lista de clientes
# La función realizará varias acciones para cada cliente/a (con bucle for).
# - Leerá el grupo de edad para definir el factor
# - Leerá el código de comportamiento para transformarlo en un valor numérico y su descripción
# - Calculará el riesgo en base a los códigos
# Retornará un diccionario de diccionarios (que se crea al principio)

def processar_clients(clients):

    clients_processats = {}

    for client in clients:

        id_client = client["id_client"]

        factor_edat = obtenir_factor_edat(client["grup_edat"])

        info_comportament = obtenir_segment_factor(client["codi_comportament"])
        segment = info_comportament["segment"]
        factors_codi = info_comportament["factor"]

        risc_final = calcular_risc(factors_codi, factor_edat)

        clients_processats[id_client] = {
            "nom": client["nom"],
            "edat": client["edat"],
            "grup_edat": client["grup_edat"],
            "codi_comportament": client["codi_comportament"],
            "segment": segment,
            "factors_codi": factors_codi,
            "factor_edat": factor_edat,
            "risc_final": risc_final
        }

    return clients_processats

In [ ]:
# COMPROBACIONES DEL PASO 4

clients_processats = processar_clients(clients)

# Vista general
print("VISTA GENERAL")
print(clients_processats)

# Ver un/a cliente/a
print("--- VEURE UN/A CLIENT/A ---")
print(clients_processats[1])

# Ver campos específicos
print("--- VEURE CAMPS ESPECÍFICS ---")
print(clients_processats[1]["nom"])
print(clients_processats[1]["segment"])
print(clients_processats[1]["risc_final"])

# Ver resumen
print("--- VEURE RESUM ---")
for id_client, dades in clients_processats.items():
    print(id_client, dades["nom"], dades["risc_final"])

VISTA GENERAL
{1: {'nom': 'Laura Serra', 'edat': 23, 'grup_edat': '18-25', 'codi_comportament': 4, 'segment': 'Cistella abandonada >24h', 'factors_codi': 0.9, 'factor_edat': 1.2, 'risc_final': 1.08}, 2: {'nom': 'Marc Vidal', 'edat': 35, 'grup_edat': '26-40', 'codi_comportament': 3, 'segment': 'Cistella abandonada <24h', 'factors_codi': 0.75, 'factor_edat': 1.0, 'risc_final': 0.75}, 3: {'nom': 'Ana López', 'edat': 29, 'grup_edat': '26-40', 'codi_comportament': 2, 'segment': 'Arriba al checkout però no paga', 'factors_codi': 0.5, 'factor_edat': 1.0, 'risc_final': 0.5}, 4: {'nom': 'Joan Riera', 'edat': 19, 'grup_edat': '18-25', 'codi_comportament': 2, 'segment': 'Arriba al checkout però no paga', 'factors_codi': 0.5, 'factor_edat': 1.2, 'risc_final': 0.6}, 5: {'nom': 'Mónica Pérez', 'edat': 47, 'grup_edat': '41-60', 'codi_comportament': 4, 'segment': 'Cistella abandonada >24h', 'factors_codi': 0.9, 'factor_edat': 0.85, 'risc_final': 0.765}, 6: {'nom': 'Luis García', 'edat': 52, 'grup_edat

In [ ]:
# PASO 5: MOSTRAR LOS DATOS

# El objetivo de la función es mostrar los resultados de processar_clients de manera legible y estructurada. La función recibe el diccionario que hemos creado de clientes procesados

from pprint import pprint

def mostrar_diccionari(clients_processats):
  pprint(clients_processats)

In [ ]:
# COMPROBACIÓN DEL PASO 5

mostrar_diccionari(clients_processats)

{1: {'codi_comportament': 4,
     'edat': 23,
     'factor_edat': 1.2,
     'factors_codi': 0.9,
     'grup_edat': '18-25',
     'nom': 'Laura Serra',
     'risc_final': 1.08,
     'segment': 'Cistella abandonada >24h'},
 2: {'codi_comportament': 3,
     'edat': 35,
     'factor_edat': 1.0,
     'factors_codi': 0.75,
     'grup_edat': '26-40',
     'nom': 'Marc Vidal',
     'risc_final': 0.75,
     'segment': 'Cistella abandonada <24h'},
 3: {'codi_comportament': 2,
     'edat': 29,
     'factor_edat': 1.0,
     'factors_codi': 0.5,
     'grup_edat': '26-40',
     'nom': 'Ana López',
     'risc_final': 0.5,
     'segment': 'Arriba al checkout però no paga'},
 4: {'codi_comportament': 2,
     'edat': 19,
     'factor_edat': 1.2,
     'factors_codi': 0.5,
     'grup_edat': '18-25',
     'nom': 'Joan Riera',
     'risc_final': 0.6,
     'segment': 'Arriba al checkout però no paga'},
 5: {'codi_comportament': 4,
     'edat': 47,
     'factor_edat': 0.85,
     'factors_codi': 0.9,
     'gru

In [ ]:
# PASO 6: PROCESAR CLIENTES

# El objetivo de la función es mostrar información básica separada por clientes
# Mostrará en pantalla esta información de cada cliente (usando bucle for) mediante prints

def mostrar_respostes(clients_processats):
    for id_client, client in clients_processats.items():
        print(f"Client: {client['nom']}")
        print(f"Edat: {client['edat']}")
        print(f"Grup d'edat: {client['grup_edat']}")
        print(f"Segment: {client['segment']}")
        print(f"Risc final: {client['risc_final']}")
        print("-" * 30)

In [ ]:
# COMPROBACIÓN DEL PASO 6

mostrar_respostes(clients_processats)

Client: Laura Serra
Edat: 23
Grup d'edat: 18-25
Segment: Cistella abandonada >24h
Risc final: 1.08
------------------------------
Client: Marc Vidal
Edat: 35
Grup d'edat: 26-40
Segment: Cistella abandonada <24h
Risc final: 0.75
------------------------------
Client: Ana López
Edat: 29
Grup d'edat: 26-40
Segment: Arriba al checkout però no paga
Risc final: 0.5
------------------------------
Client: Joan Riera
Edat: 19
Grup d'edat: 18-25
Segment: Arriba al checkout però no paga
Risc final: 0.6
------------------------------
Client: Mónica Pérez
Edat: 47
Grup d'edat: 41-60
Segment: Cistella abandonada >24h
Risc final: 0.765
------------------------------
Client: Luis García
Edat: 52
Grup d'edat: 41-60
Segment: Afegeix productes però no checkout
Risc final: 0.2125
------------------------------
Client: Pilar Sánchez
Edat: 61
Grup d'edat: 60+
Segment: Cistella abandonada <24h
Risc final: 0.5249999999999999
------------------------------
Client: Eva Martín
Edat: 38
Grup d'edat: 26-40
Segment

In [ ]:
# PASO 7 CON EXPLICACIÓN Y COMPROBACIÓN

# El objetivo de la función es orquestrar las funciones anteriores. Partirá del diccionario clients que nos ha sido dado en el enunciado
# La función procesará los clientes mediante processar_clients, mostrará las respuestsa a través de mostrar_diccionari y permetirá responder las preguntas de márqueting mediante mostrar_respostes

def main_marketing():
  clients_processats = processar_clients(clients)
  # se nutre de obtenir_segment_factor(codi_comportament),
  # obtenir_factor_edat(grup_edat)
  #calcular_risc(factor_codi, factor_edat)

  mostrar_diccionari(clients_processats)

  mostrar_respostes(clients_processats)


In [ ]:
# COMPROBACIÓN DEL PASO 7

main_marketing()

{1: {'codi_comportament': 4,
     'edat': 23,
     'factor_edat': 1.2,
     'factors_codi': 0.9,
     'grup_edat': '18-25',
     'nom': 'Laura Serra',
     'risc_final': 1.08,
     'segment': 'Cistella abandonada >24h'},
 2: {'codi_comportament': 3,
     'edat': 35,
     'factor_edat': 1.0,
     'factors_codi': 0.75,
     'grup_edat': '26-40',
     'nom': 'Marc Vidal',
     'risc_final': 0.75,
     'segment': 'Cistella abandonada <24h'},
 3: {'codi_comportament': 2,
     'edat': 29,
     'factor_edat': 1.0,
     'factors_codi': 0.5,
     'grup_edat': '26-40',
     'nom': 'Ana López',
     'risc_final': 0.5,
     'segment': 'Arriba al checkout però no paga'},
 4: {'codi_comportament': 2,
     'edat': 19,
     'factor_edat': 1.2,
     'factors_codi': 0.5,
     'grup_edat': '18-25',
     'nom': 'Joan Riera',
     'risc_final': 0.6,
     'segment': 'Arriba al checkout però no paga'},
 5: {'codi_comportament': 4,
     'edat': 47,
     'factor_edat': 0.85,
     'factors_codi': 0.9,
     'gru

### RESPUESTA A LAS PREGUNTAS DE MARKETING

**» 1. Quins clients s'han de contactar primer en una campanya de retargeting?**

Debemos responder a los clientes con un riesgo final (risc_final) más elevado, en ser los que han mostrado más intención de comprar. En este caso, se trata de:
Laura Serra, Núria Costa, Mónica Perez, Marc Vidal y Joan Riera


**» 2. Quines diferències hi ha entre clients amb el mateix comportament però d'edats diferents? (Per exemple: un client jove que abandona la cistella de compra fa 24 h té el mateix risc que un de més edat?)**

El factor edad multiplica el factor de comportamiento, por lo que la edad modifica el riesgo final.
Un/a cliente/a joven (18-25) tiene un riesgo superior al de un cliente de más edad cuando presentan el mismo comportamiento, ya que el factor de edad es más alto


**» 3. Quins clients presenten tan poc interès que no val la pena invertir-hi pressupost?**

Eva Martín tiene un código 0 (solo navega y no añade productos) por lo que muestra muy poco interés.
Luis García y Diego Romero solo han añadido productos pero no han iniciado el checjout, por lo que el interés también es bajo.


**» 4. Quin client seria el més rendible de recuperar?**

La persona más rendible de recuperar es Laura Serra


# **NIVELL 2**
## _EXERCICI 1, GESTIÓ DE RISC I PRIORITZACIÓ DE PACIENTS_

Treballes com a analista de dades en un centre mèdic especialitzat. Diversos pacients han enviat els seus símptomes a través d'un formulari online, però les dades arriben en un format desordenat i poc estructurat, i l'equip mèdic necessita que els organitzis per poder prioritzar l'ordre d'atenció segons la gravetat.
A continuació tens una llista amb informació real procedent del sistema:
                

pacients = [
    "Maria|45|tos, Fiebre Alta, dolor_pecho",
    "Luis|33|dolor_cabeza,faTiga",
    "Sara|67|fiebre alta, dificultad RESPIRAR, tos",
    "Jordi|52|fatiga , tos",
    "Anna|29|fiebre alta,dolor_cabeza"
]

               
El departament mèdic també t'ha proporcionat el següent diccionari de nivells de risc per símptoma:
                

nivells_risc = {
    "fiebre_alta": 3,
    "dificultad_respirar": 5,
    "dolor_pecho": 4,
    "tos": 1,
    "fatiga": 1,
    "dolor_cabeza": 1
}

               
🎯 Objectius
L'objectiu és transformar aquestes dades en informació útil per als metges. A partir dels registres en brut, has de:

» Convertir cada string en una estructura neta (diccionari o similar).

» Separar i normalitzar la llista de símptomes:
minúscules
eliminar espais
dividir correctament per comes

» Assignar un risc a cada símptoma utilitzant el diccionari nivells_risc.

» Calcular el risc total de cada pacient sumant els valors.

» Ordenar els pacients de major a menor risc per determinar la prioritat d'atenció.

⚠️ Important
En aquest Nivell 2 ja no et proporcionem la llista de funcions que has de crear. Ets tu qui ha de plantejar quines funcions necessites i com modularitzar el procés.
Tot i així, es recomana que:
» La funció principal orquestri tots els passos.
» Cada subfunció faci una sola tasca concreta (neteja, conversió, càlcul, ordenació...).
» La lògica estigui separada de manera neta i coherent.

✅ Resultat esperat
Un DataFrame o un diccionari amb:
nom
edat
símptomes normalitzats
risc_total
I un apartat final amb els pacients ordenats per prioritat (més risc → primer).


In [115]:
# PASO 0: INFORMACIÓN Y CREACIÓN DE DICCIONARIOS

pacients = [ "Maria|45|tos, Fiebre Alta, dolor_pecho", "Luis|33|dolor_cabeza,faTiga", "Sara|67|fiebre alta, dificultad RESPIRAR, tos", "Jordi|52|fatiga , tos", "Anna|29|fiebre alta,dolor_cabeza" ]

nivells_risc = { "fiebre_alta": 3, "dificultad_respirar": 5, "dolor_pecho": 4, "tos": 1, "fatiga": 1, "dolor_cabeza": 1 }

In [116]:
# PASO 1: Estructurar la información

# El objetivo de la función es convertir cada string en una estructura limpia, concretamente un diccionario. Recibe un string
# La función separará la info donde encuentre el caracter "|" para obtener tres campos: nombre, edad y síntomas sin procesar
# Retornará un diccionario eliminando las separaciones por "|"

def separar_datos_paciente(registro):
    nom, edat, simptomes = registro.split("|")

    return {
        "nom": nom,
        "edat": int(edat),
        "simptomes": simptomes
    }

In [117]:
# COMPROBACIÓN DEL PASO 1

for registro in pacients:
    print(separar_datos_paciente(registro))

{'nom': 'Maria', 'edat': 45, 'simptomes': 'tos, Fiebre Alta, dolor_pecho'}
{'nom': 'Luis', 'edat': 33, 'simptomes': 'dolor_cabeza,faTiga'}
{'nom': 'Sara', 'edat': 67, 'simptomes': 'fiebre alta, dificultad RESPIRAR, tos'}
{'nom': 'Jordi', 'edat': 52, 'simptomes': 'fatiga , tos'}
{'nom': 'Anna', 'edat': 29, 'simptomes': 'fiebre alta,dolor_cabeza'}


In [118]:
# PASO 2: Normalizar síntomas

# El objetivo de la función es normalizar los datos de síntomas recibidos, convirtiendo todas las letras a minúsculas, eliminando espacios y _ innecesarios y separándolos correctamente por comas
# La función usará lower, strip, replace y split para realizar esas acciones, respectivamente, normalizará el formato, y añadirá la información en una nueva lista "sintomas"
# Retornará el diccionario con el campo "sintomas" actualizado

def normalizar_sintomas(paciente):
    sintomas = paciente["simptomes"].lower().split(",")
    sintomas = [s.strip().replace(" ", "_") for s in sintomas]

    paciente["simptomes"] = sintomas

    return paciente

In [119]:
# COMPROBACIÓN DEL PASO 2

paciente = separar_datos_paciente(pacients[0])
print(paciente)

paciente = normalizar_sintomas(paciente)
print(paciente)

{'nom': 'Maria', 'edat': 45, 'simptomes': 'tos, Fiebre Alta, dolor_pecho'}
{'nom': 'Maria', 'edat': 45, 'simptomes': ['tos', 'fiebre_alta', 'dolor_pecho']}


In [120]:
# PASO 3: Cálculo de riesgo

# El objetivo de la función es calcular/asignar un nivel de riesgo a cada paciente a partir de sus síntomas y del diccionario que nos ha sido dado de nivells_risc
# La función usará un bucle para recorrer la lista de síntomas del/a paciente y obtener su valor de riesgo (mediante get(), y si un síntoma no aparece, se asignará 0). Acumulará el riesgo en el contador que hemos creado.
# Retornará el diccionario con la clave risc_total actualizada

def calcular_riesgo(paciente):
    riesgo_total = 0

    for sintoma in paciente["simptomes"]:
        riesgo_total += nivells_risc.get(sintoma, 0)

    paciente["risc_total"] = riesgo_total

    return paciente

In [122]:
# COMPROBACIÓN DEL PASO 3

paciente = separar_datos_paciente(pacients[0])
paciente = normalizar_sintomas(paciente)
paciente = calcular_riesgo(paciente)

print(paciente)

{'nom': 'Maria', 'edat': 45, 'simptomes': ['tos', 'fiebre_alta', 'dolor_pecho'], 'risc_total': 8}


In [123]:
# PASO 4: Orden por riesgo

# El objetivo de la función es ordenar los pacientes de mayor a menor riesgo. Recibe el listado de pacientes
# La función utiliza sorted en vez de sort para no modificar el listado original
# Retornará un listado ordenado

def ordenar_por_riesgo(lista_pacientes):
    return sorted(lista_pacientes,
                  key=lambda paciente: paciente["risc_total"],
                  reverse=True)


In [125]:
# COMPROBACIÓN DEL PASO 4

from pprint import pprint

pacientes_procesados = []

for registro in pacients:
    paciente = separar_datos_paciente(registro)
    paciente = normalizar_sintomas(paciente)
    paciente = calcular_riesgo(paciente)

    pacientes_procesados.append(paciente)

pprint(pacientes_procesados)

pacientes_ordenados = ordenar_por_riesgo(pacientes_procesados)

pprint(pacientes_ordenados)

[{'edat': 45,
  'nom': 'Maria',
  'risc_total': 8,
  'simptomes': ['tos', 'fiebre_alta', 'dolor_pecho']},
 {'edat': 33,
  'nom': 'Luis',
  'risc_total': 2,
  'simptomes': ['dolor_cabeza', 'fatiga']},
 {'edat': 67,
  'nom': 'Sara',
  'risc_total': 9,
  'simptomes': ['fiebre_alta', 'dificultad_respirar', 'tos']},
 {'edat': 52, 'nom': 'Jordi', 'risc_total': 2, 'simptomes': ['fatiga', 'tos']},
 {'edat': 29,
  'nom': 'Anna',
  'risc_total': 4,
  'simptomes': ['fiebre_alta', 'dolor_cabeza']}]
[{'edat': 67,
  'nom': 'Sara',
  'risc_total': 9,
  'simptomes': ['fiebre_alta', 'dificultad_respirar', 'tos']},
 {'edat': 45,
  'nom': 'Maria',
  'risc_total': 8,
  'simptomes': ['tos', 'fiebre_alta', 'dolor_pecho']},
 {'edat': 29,
  'nom': 'Anna',
  'risc_total': 4,
  'simptomes': ['fiebre_alta', 'dolor_cabeza']},
 {'edat': 33,
  'nom': 'Luis',
  'risc_total': 2,
  'simptomes': ['dolor_cabeza', 'fatiga']},
 {'edat': 52, 'nom': 'Jordi', 'risc_total': 2, 'simptomes': ['fatiga', 'tos']}]


In [126]:
# PASO 5: Main, orquestradora

# El objetivo de la función es coordinar las otras funciones que tratan los datos de los pacientes.
# En primer lugar crea una lista vacía para almacenar los pacientes ya procesados. Recorre la lista original y para cada paciente:
# Separa los datos del registro y los convierte en un diccionario, normaliza los síntomas para unificar el formato, calcula el riesgo total
# Retornará un listado ordenado segun el nivel de riesgo

def main():
    pacientes_procesados = []

    for registro in pacients:
        paciente = separar_datos_paciente(registro)
        paciente = normalizar_sintomas(paciente)
        paciente = calcular_riesgo(paciente)

        pacientes_procesados.append(paciente)

    pacientes_ordenados = ordenar_por_riesgo(pacientes_procesados)

    print("PACIENTES ORDENADOS POR PRIORIDAD\n")

    for paciente in pacientes_ordenados:
        print(paciente)

In [127]:
# COMPROBACIÓN DEL PASO 5

main()

PACIENTES ORDENADOS POR PRIORIDAD

{'nom': 'Sara', 'edat': 67, 'simptomes': ['fiebre_alta', 'dificultad_respirar', 'tos'], 'risc_total': 9}
{'nom': 'Maria', 'edat': 45, 'simptomes': ['tos', 'fiebre_alta', 'dolor_pecho'], 'risc_total': 8}
{'nom': 'Anna', 'edat': 29, 'simptomes': ['fiebre_alta', 'dolor_cabeza'], 'risc_total': 4}
{'nom': 'Luis', 'edat': 33, 'simptomes': ['dolor_cabeza', 'fatiga'], 'risc_total': 2}
{'nom': 'Jordi', 'edat': 52, 'simptomes': ['fatiga', 'tos'], 'risc_total': 2}


# **NIVELL 2**
## _EXERCICI 2, CLASSIFICACIÓ DE CLIENTS SEGURS_

Treballes com a analista de dades en una companyia asseguradora.
L'empresa vol calcular el risc dels clients, el preu estimat de la pòlissa i la probabilitat de frau, utilitzant un codi de comportament que ha estat generat per un model intern.
Tens la següent llista de clients:
                
clients = [
    {"nom": "Ana", "edat": 42, "codi": 0},
    {"nom": "Carlos", "edat": 30, "codi": 2},
    {"nom": "Isabel", "edat": 55, "codi": 3},
    {"nom": "Jorge", "edat": 40, "codi": 1},
    {"nom": "Marta", "edat": 28, "codi": 0}
]
                       
I aquest diccionari amb les categories del sistema:
                
nivells_seguro = {
    0: {"categoria": "Baix risc", "preu_base": 120, "fraude": 0.01},
    1: {"categoria": "Risc mitjà", "preu_base": 200, "fraude": 0.05},
    2: {"categoria": "Risc alt", "preu_base": 350, "fraude": 0.15},
    3: {"categoria": "Risc crític", "preu_base": 500, "fraude": 0.30}
}
                 
      
** El preu final s'ha de calcular amb la fórmula:**
preu_final = preu_base * (1 + edat/100)

**🎯 Objectius**
» Classificar cada client utilitzant nivells_seguro (sense if/elif).
» Afegir:
La categoria
El preu base
La probabilitat de frau
» Calcular el preu final de la pòlissa amb la fórmula donada.
» Crear un DataFrame o diccionari amb la informació final.
» Ordenar els clients:
Per probabilitat de frau (descendent).
Per preu final.
 Important
Igual que a l'Exercici 1, en aquest nivell ets tu qui ha de definir les funcions necessàries. Es recomana mantenir el codi modular, on:
» Cada funció resol un pas del procés.
» La funció principal delega totes les tasques.

✅ Resultat esperat
Un DataFrame o un diccionari amb:
Nom
Edat
Categoria
Preu_base
Preu_final
Probabilitat_frau
Un apartat d'anàlisi on s'indiqui:
Quin client té més probabilitat de frau
Quin té el preu final més alt
Qui seria prioritari de revisar manualment


In [128]:
# PASO 0: INFORMACIÓN Y CREACIÓN DE DICCIONARIOS

clients = [ {"nom": "Ana", "edat": 42, "codi": 0}, {"nom": "Carlos", "edat": 30, "codi": 2}, {"nom": "Isabel", "edat": 55, "codi": 3}, {"nom": "Jorge", "edat": 40, "codi": 1}, {"nom": "Marta", "edat": 28, "codi": 0} ]

nivells_seguro = { 0: {"categoria": "Baix risc", "preu_base": 120, "fraude": 0.01}, 1: {"categoria": "Risc mitjà", "preu_base": 200, "fraude": 0.05}, 2: {"categoria": "Risc alt", "preu_base": 350, "fraude": 0.15}, 3: {"categoria": "Risc crític", "preu_base": 500, "fraude": 0.30} }

In [129]:
# PASO 1: Clasificar clientes

# El objetivo de la función es devolver información sobre los clientes en base a su código de comportamiento, que encontramos en el diccionario nivells_seguro. Recibe un cliente de la lista y el diccionario dado.
# La función crea nuevas variables en los clientes, concretamente: categoria, preu_base, fraude
# La función retorna el diccionario actualizado

def afegir_cat_pb_f(client, nivells_seguro):
    nivell_seguro = nivells_seguro[client["codi"]]

    client["categoria"] = nivell_seguro["categoria"]
    client["preu_base"] = nivell_seguro["preu_base"]
    client["fraude"] = nivell_seguro["fraude"]

    return client

In [130]:
# COMPROBACIÓN DEL PASO 1

cliente_prueba = {"nom": "Natalia", "edat": 30, "codi": 0}

resultado = afegir_cat_pb_f(cliente_prueba, nivells_seguro)

print(resultado)

{'nom': 'Natalia', 'edat': 30, 'codi': 0, 'categoria': 'Baix risc', 'preu_base': 120, 'fraude': 0.01}


In [131]:
# PASO 2: Calcular el precio final

# El objetivo de la función es calcular el precio final. Recibe el precio base y la edad
# La función realizará la siguiente operación: preu_final = preu_base * (1 + edat/100)
# Retornará el resultado del cálculo

def calcular_preu_final(preu_base, edat):

  preu_final = round(preu_base * (1 + edat/100), 2)

  return preu_final

In [132]:
# COMPROBACIÓN DEL PASO 2

calcular_preu_final(120, 42)

170.4

In [133]:
# PASO 3, Añadir precio final

# El objetivo de la función es añadir el precio final a la información del cliente. Recibe un cliente
# La función añade una nuvea clave al diccionario
# Retornará el cliente enriquecido

def afegir_preu_final(client):
    client["preu_final"] = calcular_preu_final(client["preu_base"], client["edat"])

    return client

In [134]:
# COMPROBACIÓN DEL PASO 3

cliente_prueba = {
    "nom": "Carlos",
    "edat": 30,
    "codi": 2,
    "categoria": "Risc alt",
    "preu_base": 350,
    "fraude": 0.15
}

resultado = afegir_preu_final(cliente_prueba)

print(resultado)

{'nom': 'Carlos', 'edat': 30, 'codi': 2, 'categoria': 'Risc alt', 'preu_base': 350, 'fraude': 0.15, 'preu_final': 455.0}


In [135]:
# PASO 4: Crear diccionario con la información actualizada

# El objetivo de la función es crear un diccionario con todos los datos de todos los clientes actualizados. Recibe una lista
# Mediante un bucle for, recorre cada cliente para añadirlo todo en una nueva lista
# Retorna un diccionario actualizado

def processar_clients():
    clients_processats = {}

    for client in clients:
        afegir_cat_pb_f(client, nivells_seguro)
        afegir_preu_final(client)

        clients_processats[client["nom"]] = client

    return clients_processats

In [136]:
# COMPROBACIÓN DEL PASO 4

processar_clients()

{'Ana': {'nom': 'Ana',
  'edat': 42,
  'codi': 0,
  'categoria': 'Baix risc',
  'preu_base': 120,
  'fraude': 0.01,
  'preu_final': 170.4},
 'Carlos': {'nom': 'Carlos',
  'edat': 30,
  'codi': 2,
  'categoria': 'Risc alt',
  'preu_base': 350,
  'fraude': 0.15,
  'preu_final': 455.0},
 'Isabel': {'nom': 'Isabel',
  'edat': 55,
  'codi': 3,
  'categoria': 'Risc crític',
  'preu_base': 500,
  'fraude': 0.3,
  'preu_final': 775.0},
 'Jorge': {'nom': 'Jorge',
  'edat': 40,
  'codi': 1,
  'categoria': 'Risc mitjà',
  'preu_base': 200,
  'fraude': 0.05,
  'preu_final': 280.0},
 'Marta': {'nom': 'Marta',
  'edat': 28,
  'codi': 0,
  'categoria': 'Baix risc',
  'preu_base': 120,
  'fraude': 0.01,
  'preu_final': 153.6}}

In [137]:
# PASO 5, ordenar clientes

# El objetivo de la función es ordenar los clientes en base a la probabilidad de fraude y el precio. Recibe el diccionario de clientes ya procesados
# La función utilizará sorted para la ordenación
# Devolverá una lista ordenada

def ordenar_clients(clients_processats):

    ordenats = sorted(
        clients_processats.items(),
        key=lambda x: (x[1]["fraude"], x[1]["preu_final"]),
        reverse=True
    )
    return ordenats

In [138]:
# COMPROBACIÓN DEL PASO 5

from pprint import pprint

clients_processats = processar_clients()

clients_ordenats = ordenar_clients(clients_processats)

pprint(clients_ordenats)

[('Isabel',
  {'categoria': 'Risc crític',
   'codi': 3,
   'edat': 55,
   'fraude': 0.3,
   'nom': 'Isabel',
   'preu_base': 500,
   'preu_final': 775.0}),
 ('Carlos',
  {'categoria': 'Risc alt',
   'codi': 2,
   'edat': 30,
   'fraude': 0.15,
   'nom': 'Carlos',
   'preu_base': 350,
   'preu_final': 455.0}),
 ('Jorge',
  {'categoria': 'Risc mitjà',
   'codi': 1,
   'edat': 40,
   'fraude': 0.05,
   'nom': 'Jorge',
   'preu_base': 200,
   'preu_final': 280.0}),
 ('Ana',
  {'categoria': 'Baix risc',
   'codi': 0,
   'edat': 42,
   'fraude': 0.01,
   'nom': 'Ana',
   'preu_base': 120,
   'preu_final': 170.4}),
 ('Marta',
  {'categoria': 'Baix risc',
   'codi': 0,
   'edat': 28,
   'fraude': 0.01,
   'nom': 'Marta',
   'preu_base': 120,
   'preu_final': 153.6})]


In [139]:
# PASO 6: Main, orquestradora

# El objetivo de la función es coordinar los pasos anteriores.
# La función obtiene los clientes procesados, los ordena y muestra los resultados
# Devolverá un diccionario ordenado segun los criterios fraude y preu_base

def main_2():
    clients_processats = processar_clients()

    clients_ordenats = ordenar_clients(clients_processats)

    return clients_ordenats

In [ ]:
# COMPROBACIÓN DEL PASO 6

from pprint import pprint

resultat_final = main_2()

pprint(resultat_final)

[('Isabel',
  {'categoria': 'Risc crític',
   'codi': 3,
   'edat': 55,
   'fraude': 0.3,
   'nom': 'Isabel',
   'preu_base': 500,
   'preu_final': 775.0}),
 ('Carlos',
  {'categoria': 'Risc alt',
   'codi': 2,
   'edat': 30,
   'fraude': 0.15,
   'nom': 'Carlos',
   'preu_base': 350,
   'preu_final': 455.0}),
 ('Jorge',
  {'categoria': 'Risc mitjà',
   'codi': 1,
   'edat': 40,
   'fraude': 0.05,
   'nom': 'Jorge',
   'preu_base': 200,
   'preu_final': 280.0}),
 ('Ana',
  {'categoria': 'Baix risc',
   'codi': 0,
   'edat': 42,
   'fraude': 0.01,
   'nom': 'Ana',
   'preu_base': 120,
   'preu_final': 170.4}),
 ('Marta',
  {'categoria': 'Baix risc',
   'codi': 0,
   'edat': 28,
   'fraude': 0.01,
   'nom': 'Marta',
   'preu_base': 120,
   'preu_final': 153.6})]


### ANÀLISI

**» 1. Quin client té més probabilitat de frau?**

Isabel, amb un risc crític (0.3)

**» 2. Quin té el preu final més alt?**

Isabel, amb 775€

**» 2. Qui seria prioritari de revisar manualment?**

Isabel, en ser qui més risc de frau té i qui té un preu final més elevat

# **NIVELL 3**
## _EXERCICI 1, GENERADOR DE CONTRASSENYES SEGURES_

Ets analista de dades dins el departament de Ciberseguretat d'una empresa tecnològica. S'ha detectat que la majoria de treballadors continuen utilitzant contrasenyes molt dèbils com “Asdf1234”, dates d'aniversari o combinacions fàcils d'endevinar. Per aquest motiu, el responsable del departament t'ha demanat desenvolupar un generador de contrasenyes segures, totalment parametritzable i modular.

**📝 Requisits del generador**
Has de crear una funció principal i totes les subfuncions necessàries per:
» Generar contrasenyes de longitud variable.
» Decidir si han d'incloure:
Majúscules
Minúscules
Números
Signes especials
» Assegurar-te que com a mínim apareix un caràcter de cada tipus seleccionat.
» Aleatoritzar completament la contrasenya.
» Retornar la contrasenya generada.

**📍 Nota important **
Explora el funcionament del mòdul random de la llibreria numpy

La funció principal del generador ha de tenir valors per defecte, exactament els següents:
majuscules=True
minuscules=True
numeros=True
signes=False

Aquests paràmetres són obligatoris i formen part del disseny del sistema.
👁️ Exemple d'ús esperat
Output possible:


In [144]:
# PASO 1: Obtenir paràmetres permesos

# En esta función se determinan los valores permitidos. Recibe los parámetros que queremos incluir.
# La función retornará una cadena con los valores selecciondos

def obtenir_caracters_permesos(majuscules, minuscules, numeros, signes):
    caracters = ""

    if majuscules:
        caracters += "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    if minuscules:
        caracters += "abcdefghijklmnopqrstuvwxyz"

    if numeros:
        caracters += "0123456789"

    if signes:
        caracters += "!@#$"

    return caracters

In [145]:
# COMPROBACIONES DEL PASO 1

resultado = obtenir_caracters_permesos(True, True, True, True)
print(f"Primera combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, False, True, True)
print(f"Segunda combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, True, False, True)
print(f"Tercera combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, True, True, False)
print(f"Cuarta combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, False, True, True)
print(f"Quinta combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, True, False, True)
print(f"Sexta combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, True, True, False)
print(f"Séptima combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, False, False, True)
print(f"Octava combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, False, True, False)
print(f"Novena combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, True, False, False)
print(f"Décima combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, False, False, True)
print(f"Decimoprimera combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, False, True, False)
print(f"Decimosegunda combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, True, False, False)
print(f"Decimotercera combinación: {resultado}")

resultado = obtenir_caracters_permesos(True, False, False, False)
print(f"Decimocuarta combinación: {resultado}")

resultado = obtenir_caracters_permesos(False, False, False, False)
print(f"Decimoquinta combinación: {resultado}")

Primera combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$
Segunda combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$
Tercera combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz!@#$
Cuarta combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789
Quinta combinación: 0123456789!@#$
Sexta combinación: abcdefghijklmnopqrstuvwxyz!@#$
Séptima combinación: abcdefghijklmnopqrstuvwxyz0123456789
Octava combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$
Novena combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789
Décima combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz
Decimoprimera combinación: !@#$
Decimosegunda combinación: 0123456789
Decimotercera combinación: abcdefghijklmnopqrstuvwxyz
Decimocuarta combinación: ABCDEFGHIJKLMNOPQRSTUVWXYZ
Decimoquinta combinación: 


In [153]:
# PASO 2: Afegir caràcters obligatoris

# El objetivo de la función es hacer que cumpla los criterios. Recibe los criterios que tendrá la contraseña
# La función creará una lista donde añadirá un caracter de cada grupo necesario
# Retornará una lista con los valores

import numpy as np

def añadir_caracteres_obligatorios(majuscules, minuscules, numeros, signes):

  valors = []

  if majuscules is True:
    valors.append(str(np.random.choice(list("ABCDEFGHIJKLMNOPQRSTUVWXYZ"))))

  if minuscules is True:
    valors.append(str(np.random.choice(list("abcdefghijklmnopqrstuvwxyz"))))

  if numeros is True:
    valors.append(str(np.random.choice(list("0123456789"))))

  if signes is True:
    valors.append(str(np.random.choice(list("!@#$"))))

  return valors

In [154]:
# COMPROBACIONES DEL PASO 2

resultado = añadir_caracteres_obligatorios(True, True, True, True)
print(f"Primera combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, False, True, True)
print(f"Segunda combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, True, False, True)
print(f"Tercera combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, True, True, False)
print(f"Cuarta combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, False, True, True)
print(f"Quinta combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, True, False, True)
print(f"Sexta combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, True, True, False)
print(f"Séptima combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, False, False, True)
print(f"Octava combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, False, True, False)
print(f"Novena combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, True, False, False)
print(f"Décima combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, False, False, True)
print(f"Decimoprimera combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, False, True, False)
print(f"Decimosegunda combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, True, False, False)
print(f"Decimotercera combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(True, False, False, False)
print(f"Decimocuarta combinación: {resultado}")

resultado = añadir_caracteres_obligatorios(False, False, False, False)
print(f"Decimoquinta combinación: {resultado}")

Primera combinación: ['Y', 'o', '5', '#']
Segunda combinación: ['I', '3', '!']
Tercera combinación: ['P', 'c', '@']
Cuarta combinación: ['V', 't', '3']
Quinta combinación: ['1', '$']
Sexta combinación: ['g', '#']
Séptima combinación: ['j', '8']
Octava combinación: ['T', '$']
Novena combinación: ['L', '6']
Décima combinación: ['E', 'w']
Decimoprimera combinación: ['$']
Decimosegunda combinación: ['1']
Decimotercera combinación: ['s']
Decimocuarta combinación: ['L']
Decimoquinta combinación: []


In [158]:
# PASO 3: Omplir contrassenya

# El objetivo de esta función es "rellenar" la contrasenya para llegar al nº de valores solicitados (longitud)
# Recibe la lista de carácteres obligatorios (valors), comprueba su longitud. A través de bucle while, mientras no se llegue a la longitud deseada, se escoge un carácter y se añade a la lista
# Devuelve la lista completa
# Opto por no modificar la lista original

def omplir_contrassenya(valors, caracters, longitud):
    completa = valors.copy()

    while len(completa) < longitud:
        caracter = str(np.random.choice(list(caracters)))
        completa.append(caracter)

    return completa

In [160]:
# COMPROBACIÓN DEL PASO 3

caracters = obtenir_caracters_permesos(
    majuscules=True,
    minuscules=True,
    numeros=True,
    signes=False
)

valors = añadir_caracteres_obligatorios(
    majuscules=True,
    minuscules=True,
    numeros=True,
    signes=False
)

resultado = omplir_contrassenya(valors, caracters, 10)

print(resultado)

['Y', 's', '4', 'a', '4', 'O', 'G', 'x', '8', '3']


In [161]:
# PASO 4: Barrejar valors

# El objetivo de la función es mezclar aleatoriamente los valores de la lista
# Recibe la lista, mediante shuffle se mezcla, devolverá una lista
# Opto por no modificar la lista original

import numpy as np

def barrejar_valors(completa):

    barrejada = completa.copy()

    np.random.shuffle(barrejada)

    return barrejada

In [167]:
# COMPROBACIÓN DEL PASO 4

llista = ['A', 'm', '7', 'x', 'Q', '2', '@', '?', 'n', 'J']

resultat = barrejar_valors(llista)

print("Llista original:", llista)
print("Llista barrejada:", resultat)

Llista original: ['A', 'm', '7', 'x', 'Q', '2', '@', '?', 'n', 'J']
Llista barrejada: ['2', 'x', 'n', 'Q', '7', '?', 'A', '@', 'm', 'J']


In [169]:
# PASO 5: Validación de la contraseña

# El objetivo de la función es validar que se cumplen los criterios de longitud y de un símbolo de cada. Recibe la lista (llista barrejada) y la longitud
# La función recorrerá cada carácter para comprobar si se cumple o no con un bucle for
# Si es correcto, retornará un mensaje de validación, si es incorrecto, uno de error

def validar(barrejada, longitud):
    if len(barrejada) < longitud:
        print(f"La contrassenya ha de tenir almenys {longitud} caràcters")
        return False

    tiene_mayuscula = False
    tiene_minuscula = False
    tiene_numero = False
    tiene_simbolo = False

    for caracter in barrejada:
        if caracter.isupper():
            tiene_mayuscula = True
        elif caracter.islower():
            tiene_minuscula = True
        elif caracter.isdigit():
            tiene_numero = True
        else:
            tiene_simbolo = True

    if not tiene_mayuscula:
        print("Falta una mayúscula")
    if not tiene_minuscula:
        print("Falta una minúscula")
    if not tiene_numero:
        print("Falta un número")
    if not tiene_simbolo:
        print("Falta un símbolo")

    if (tiene_mayuscula and tiene_minuscula and
            tiene_numero and tiene_simbolo):
        print("La contraseña es válida")

    return (
        tiene_mayuscula and
        tiene_minuscula and
        tiene_numero and
        tiene_simbolo
    )

In [170]:
# COMPROBACIONES DEL PASO 5

barrejada = ['x', 'J', 'm', '@', 'Q', 'n', '7', '?', '2', 'A']
validar(barrejada, 10)

print()

barrejada = ['x', 'J', 'm', '@', 'Q', 'n', '7', '?', '2'] # 9 valores
validar(barrejada, 10)

print()

barrejada = ['x', 'j', 'm', '@', 'q', 'n', '7', '?', '2', 'a'] # sin mayúsculas
validar(barrejada, 10)

print()

barrejada = ['X', 'J', 'M', '@', 'Q', 'N', '7', '?', '2', 'A'] # sin minúsculas
validar(barrejada, 10)

print()

barrejada = ['x', 'J', 'm', '@', 'Q', 'n', 'u', '?', 'n', 'A'] # sin números
validar(barrejada, 10)

print()

barrejada = ['x', 'J', 'm', '3', 'Q', 'n', '7', 'm', '2', 'A'] # sin símbolos
validar(barrejada, 10)

print()

barrejada = ['x', 'J', 'm', 't', 'Q', 'n', 'U', 'm', 'N', 'A'] # dos criterios a la vez (sin símbolos ni números)
validar(barrejada, 10)

print()

La contraseña es válida

La contrassenya ha de tenir almenys 10 caràcters

Falta una mayúscula

Falta una minúscula

Falta un número

Falta un símbolo

Falta un número
Falta un símbolo



In [172]:
# PASO 6: Main, orquestradora, generar contrassenya

# El objetivo de la función es coordinar todos los pasos anteriores
# La función define los carácteres permitidos, añade un carácter obligatorio de cada tipo, rellena la contraseña hasta la longitud deseada, mezcla esa lista y valida que se cumplan los criterios
# La función retorna un listado y su validación

import numpy as np

def generar_contrassenya():
    longitud = 10
    majuscules = True
    minuscules = True
    numeros = True
    signes = True

    caracters = obtenir_caracters_permesos(
        majuscules, minuscules, numeros, signes
    )

    valors = añadir_caracteres_obligatorios(
        majuscules, minuscules, numeros, signes
    )

    completa = omplir_contrassenya(
        valors, caracters, longitud
    )

    barrejada = barrejar_valors(completa)

    contrassenya = "".join(barrejada)

    print("Contrasenya generada:", contrassenya)

    validar(contrassenya, longitud)

In [176]:
# COMPROBACIÓN DEL PASO 6

generar_contrassenya()
print()
generar_contrassenya()

Contrasenya generada: !0lB8ihkZ0
La contraseña es válida

Contrasenya generada: F#rC5r1VxU
La contraseña es válida


# **NIVELL 3**
## _EXERCICI 2, PROCESSAMENT AUTOMÀTIC DE DADES ESPORTIVES_

Ara formes part d'un equip d'analítica dins una organització esportiva que gestiona dades de la lliga catalana. La teva companya de feina vol automatitzar la lectura del seu arxiu històric de resultats i extreure les estadístiques més importants. T'ha demanat fer un petit sistema en Python que processi el fitxer 📄 "historic partits.txt"

**🎯 Objectius**
Mitjançant un codi modular, clar i documentat, el teu programa ha de calcular:

**» 1.Quin equip és el més golejador.**

**» 2.Quin és el més golejat.**

**» 3.El nombre total de gols que ha fet cada equip.**

**» 4.La classificació final ordenada per punts on cada victòria suma 3 pts, cada empat 1 pts y cada derrota 0 pts.**

En aquest nivell no es dona cap funció. Ets tu qui ha de dissenyar un flux complet. El codi ha de reflectir una bona organització interna.

**✅ Resultat esperat**
Un DataFrame o diccionari final semblant a:

A més del text:
Equip més golejador: _______
Equip més golejat: ________
Classificació final:
1. _______
2. _______
...


In [310]:
# PASO 0: Importar y leer archivo

# Leemos el archivo línea por línea mediante bucle for
# Muestra el contenido en pantalla

def lectura():
  with open("historic partits.txt", "r", encoding="utf-8") as archivo:
    lineas = []
    for linea in archivo:
        lineas.append(linea)
    return lineas

In [311]:
# COMPROBACIÓN

from pprint import pprint

pprint(lineas[:3])

['Manlleu\t0-1\tGranollers\n',
 'Vilafranca\t4-5\tTerrassa\n',
 'Olot\t5-0\tGranollers\n']


In [312]:
# PASO 1: Limpiar formato

 # El objetivo de la función es "limpiar" el formato, consiguiendo que cada línea sea un string sin espacios ni tabulaciones ("\t")

def preparar_partidos(lineas):
    partidos = []

    for linea in lineas:
        linea = linea.strip()

        datos = linea.split("\t")

        local = datos[0]
        resultado = datos[1]
        visitante = datos[2]

        goles_local, goles_visitante = resultado.split("-")

        partidos.append([
            local,
            int(goles_local),
            int(goles_visitante),
            visitante
        ])

    return partidos

In [313]:
# COMPROBACIÓN PASO 1

preparar_partidos(lineas)

[['Manlleu', 0, 1, 'Granollers'],
 ['Vilafranca', 4, 5, 'Terrassa'],
 ['Olot', 5, 0, 'Granollers'],
 ['Vilafranca', 5, 5, 'Manlleu'],
 ['Cornellà', 5, 1, 'Prat'],
 ['Cerdanyola', 0, 5, 'Vilafranca'],
 ['Sant Andreu', 2, 3, 'Cornellà'],
 ['Vilafranca', 5, 4, 'Olot'],
 ['FC Barcelona', 4, 4, 'Badalona'],
 ['Granollers', 1, 5, 'Sabadell'],
 ['Granollers', 0, 5, 'Europa'],
 ['Prat', 5, 2, 'Llagostera'],
 ['Llagostera', 2, 5, 'Vilafranca'],
 ['Sant Andreu', 1, 4, 'Figueres'],
 ['Figueres', 1, 2, 'Cerdanyola'],
 ['Granollers', 0, 4, 'Cornellà'],
 ['Figueres', 2, 3, 'RCD Espanyol'],
 ['Llagostera', 2, 5, 'Figueres'],
 ['Cornellà', 3, 1, 'Reus Deportiu'],
 ['Europa', 5, 1, 'Granollers'],
 ['Terrassa', 0, 4, 'Europa'],
 ['Olot', 1, 2, 'Sant Andreu'],
 ['Vilafranca', 5, 2, 'Europa'],
 ['Vilafranca', 1, 5, 'Reus Deportiu'],
 ['Reus Deportiu', 4, 4, 'Nàstic de Tarragona'],
 ['Prat', 3, 0, 'Lleida Esportiu'],
 ['Sant Andreu', 1, 2, 'Cerdanyola'],
 ['Granollers', 2, 0, 'Prat'],
 ['Vilafranca', 5, 5,

In [314]:
# PASO 2: Crear diccionario

# El objetivo de la función es crear los distintos equipos donde más tarde se haran los recuentos. Recibe la lista de partidos
# Mediante bucle for, pediremos la posición 1 ([0]), referente al equipo local y la 4 ([3]) referente al equipo visitante para consultar si los equipos que van apareciendo ya "tienen una ficha creada" en nuestro diccionario o todavía no
# Devolverá el diccionario actualizado

def crear_equipos(partidos):
    equipos = {}

    for partido in partidos:
        local = partido[0]
        visitante = partido[3]

        if local not in equipos:
            equipos[local] = {
                "GF": 0,
                "GC": 0,
                "Puntos": 0
            }

        if visitante not in equipos:
            equipos[visitante] = {
                "GF": 0,
                "GC": 0,
                "Puntos": 0
            }

    return equipos

In [315]:
# COMPROBACIÓN DEL PASO 2

partidos = preparar_partidos(lineas)

crear_equipos(partidos)

{'Manlleu': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Granollers': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Vilafranca': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Terrassa': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Olot': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Cornellà': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Prat': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Cerdanyola': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Sant Andreu': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'FC Barcelona': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Badalona': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Sabadell': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Europa': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Llagostera': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Figueres': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'RCD Espanyol': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Reus Deportiu': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Nàstic de Tarragona': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Lleida Esportiu': {'GF': 0, 'GC': 0, 'Puntos': 0},
 'Girona FC': {'GF': 0, 'GC': 0, 'Puntos': 0}}

In [344]:
# PASO 3: Procesar goles

# El objetivo de la función es sacar la información de las líneas para añadirlas al diccionario correspondiente, primero centrándonos solo en los goles. Recbie el listado de partidos
# Mediante un bucle for, recorrerá cada linea para añadir los goles que ha marcado cada equipo y completar así ambos campos (goles a favor y goles en contra).
# La función retornará el diccionario actualizado

def procesar_goles(partidos):

  equipos = crear_equipos(partidos)

  for partido in partidos:
    local = partido[0]
    visitante = partido[3]

    goles_local = partido[1]
    goles_visitante = partido[2]

    equipos[local]["GF"] += goles_local
    equipos[local]["GC"] += goles_visitante
    equipos[visitante]["GF"] += goles_visitante
    equipos[visitante]["GC"] += goles_local

  return equipos

In [345]:
# COMPROBACIÓN DEL PASO 3

from pprint import pprint

equipos = procesar_goles(partidos)

pprint(equipos)

{'Badalona': {'GC': 124, 'GF': 134, 'Puntos': 0},
 'Cerdanyola': {'GC': 133, 'GF': 117, 'Puntos': 0},
 'Cornellà': {'GC': 146, 'GF': 147, 'Puntos': 0},
 'Europa': {'GC': 93, 'GF': 93, 'Puntos': 0},
 'FC Barcelona': {'GC': 115, 'GF': 125, 'Puntos': 0},
 'Figueres': {'GC': 154, 'GF': 161, 'Puntos': 0},
 'Girona FC': {'GC': 94, 'GF': 139, 'Puntos': 0},
 'Granollers': {'GC': 117, 'GF': 122, 'Puntos': 0},
 'Llagostera': {'GC': 142, 'GF': 159, 'Puntos': 0},
 'Lleida Esportiu': {'GC': 133, 'GF': 129, 'Puntos': 0},
 'Manlleu': {'GC': 118, 'GF': 106, 'Puntos': 0},
 'Nàstic de Tarragona': {'GC': 148, 'GF': 130, 'Puntos': 0},
 'Olot': {'GC': 144, 'GF': 132, 'Puntos': 0},
 'Prat': {'GC': 111, 'GF': 106, 'Puntos': 0},
 'RCD Espanyol': {'GC': 144, 'GF': 131, 'Puntos': 0},
 'Reus Deportiu': {'GC': 111, 'GF': 119, 'Puntos': 0},
 'Sabadell': {'GC': 121, 'GF': 141, 'Puntos': 0},
 'Sant Andreu': {'GC': 134, 'GF': 123, 'Puntos': 0},
 'Terrassa': {'GC': 164, 'GF': 147, 'Puntos': 0},
 'Vilafranca': {'GC': 1

In [346]:
# PASO 4: Procesar la puntuación

# El objetivo de la función es sacar la información de las líneas para añadirlas al diccionario correspondiente, ahora centrándonos en los puntos. Recbie el listado de partidos y equipos
# Mediante un bucle for, recorrerá cada linea para contabilizar los goles y actualizar el campo de puntos
# La función retornará el diccionario actualizado

def procesar_puntos(partidos, equipos):

    for partido in partidos:

        local = partido[0]
        goles_local = partido[1]
        goles_visitante = partido[2]
        visitante = partido[3]

        if goles_local > goles_visitante:
            equipos[local]["Puntos"] += 3

        elif goles_local < goles_visitante:
            equipos[visitante]["Puntos"] += 3

        else:
            equipos[local]["Puntos"] += 1
            equipos[visitante]["Puntos"] += 1

    return equipos

In [351]:
# COMPROBACIÓN DEL PASO 4

from pprint import pprint

equipos = crear_equipos(partidos)

equipos = procesar_goles(partidos)

equipos = procesar_puntos(partidos, equipos)

pprint(equipos)

{'Badalona': {'GC': 124, 'GF': 134, 'Puntos': 68},
 'Cerdanyola': {'GC': 133, 'GF': 117, 'Puntos': 56},
 'Cornellà': {'GC': 146, 'GF': 147, 'Puntos': 82},
 'Europa': {'GC': 93, 'GF': 93, 'Puntos': 41},
 'FC Barcelona': {'GC': 115, 'GF': 125, 'Puntos': 73},
 'Figueres': {'GC': 154, 'GF': 161, 'Puntos': 79},
 'Girona FC': {'GC': 94, 'GF': 139, 'Puntos': 96},
 'Granollers': {'GC': 117, 'GF': 122, 'Puntos': 67},
 'Llagostera': {'GC': 142, 'GF': 159, 'Puntos': 94},
 'Lleida Esportiu': {'GC': 133, 'GF': 129, 'Puntos': 75},
 'Manlleu': {'GC': 118, 'GF': 106, 'Puntos': 57},
 'Nàstic de Tarragona': {'GC': 148, 'GF': 130, 'Puntos': 68},
 'Olot': {'GC': 144, 'GF': 132, 'Puntos': 64},
 'Prat': {'GC': 111, 'GF': 106, 'Puntos': 56},
 'RCD Espanyol': {'GC': 144, 'GF': 131, 'Puntos': 80},
 'Reus Deportiu': {'GC': 111, 'GF': 119, 'Puntos': 68},
 'Sabadell': {'GC': 121, 'GF': 141, 'Puntos': 85},
 'Sant Andreu': {'GC': 134, 'GF': 123, 'Puntos': 62},
 'Terrassa': {'GC': 164, 'GF': 147, 'Puntos': 74},
 'Vi

In [352]:
# PASO 5: Buscar al equipo más goleador

# El objetivo de la función es buscar el equipo que más goles haya anotado. Recibe el diccionario de equipos
# Se crea un contador a 0 y, a través de un bucle for, vamos recorriendo cada equipo para quedarnos con el número más grande
# Retornará el equipo y el contador

def equipo_mas_goleador(equipos):
    max_goles_f = 0
    equipo = ""

    for nombre, datos in equipos.items():
        if datos["GF"] > max_goles_f:
            max_goles_f = datos["GF"]
            equipo = nombre

    return equipo, max_goles_f

In [353]:
# COMPROBACIÓN DEL PASO 5

mas_goleador = equipo_mas_goleador(equipos)

print(f"El equipo más goleador es el {mas_goleador[0]}, con un total de {mas_goleador[1]} goles")

El equipo más goleador es el Figueres, con un total de 161 goles


In [354]:
# PASO 6: Buscar al equipo más goleado

# El objetivo de la función es buscar el equipo que más goles haya recibido. Recibe el diccionario de equipos
# Se crea un contador a 0 y, a través de un bucle for, vamos recorriendo cada equipo para quedarnos con el número más grande
# Retornará el equipo y el contador

def equipo_mas_goleado(equipos):
    max_goles_c = 0
    equipo = ""

    for nombre, datos in equipos.items():
        if datos["GC"] > max_goles_c:
            max_goles_c = datos["GC"]
            equipo = nombre

    return equipo, max_goles_c

In [356]:
# COMPROBACIÓN DEL PASO 6

mas_goleado = equipo_mas_goleado(equipos)

print(f"El equipo más goleado es el {mas_goleado[0]}, con un total de {mas_goleado[1]} goles")

El equipo más goleado es el Vilafranca, con un total de 172 goles


In [357]:
# PASO 7: Contador de goles por equipo

# El objetivo de la función es contar los goles a favor de cada equipo

def ordenar_por_puntos(equipos):
    return sorted(equipos.items(), key=lambda x: x[1]["Puntos"], reverse=True)


In [361]:
# COMPROBACIÓN PASO 7

ranking_goles = ordenar_por_puntos(equipos)

for equipo, datos in ranking_goles:
    print(equipo, datos["Puntos"])

Girona FC 96
Llagostera 94
Sabadell 85
Cornellà 82
RCD Espanyol 80
Figueres 79
Lleida Esportiu 75
Terrassa 74
FC Barcelona 73
Vilafranca 72
Badalona 68
Reus Deportiu 68
Nàstic de Tarragona 68
Granollers 67
Olot 64
Sant Andreu 62
Manlleu 57
Prat 56
Cerdanyola 56
Europa 41


In [363]:
# PASO 8: Ordenar los resultados

# El objetivo de la función es ordenar los equipos que recibe (usando sorted para no modificar el original) en base a los puntos que gana
# Retornará

def ordenar_clasificacion(equipos):
    clasificacion = sorted(
        equipos.items(),
        key=lambda x: x[1]["Puntos"],
        reverse=True
    )

    return clasificacion

In [368]:
# COMPROBACIÓN DEL PASO 8

ordenar_clasificacion(equipos)

[('Girona FC', {'GF': 139, 'GC': 94, 'Puntos': 96}),
 ('Llagostera', {'GF': 159, 'GC': 142, 'Puntos': 94}),
 ('Sabadell', {'GF': 141, 'GC': 121, 'Puntos': 85}),
 ('Cornellà', {'GF': 147, 'GC': 146, 'Puntos': 82}),
 ('RCD Espanyol', {'GF': 131, 'GC': 144, 'Puntos': 80}),
 ('Figueres', {'GF': 161, 'GC': 154, 'Puntos': 79}),
 ('Lleida Esportiu', {'GF': 129, 'GC': 133, 'Puntos': 75}),
 ('Terrassa', {'GF': 147, 'GC': 164, 'Puntos': 74}),
 ('FC Barcelona', {'GF': 125, 'GC': 115, 'Puntos': 73}),
 ('Vilafranca', {'GF': 157, 'GC': 172, 'Puntos': 72}),
 ('Badalona', {'GF': 134, 'GC': 124, 'Puntos': 68}),
 ('Reus Deportiu', {'GF': 119, 'GC': 111, 'Puntos': 68}),
 ('Nàstic de Tarragona', {'GF': 130, 'GC': 148, 'Puntos': 68}),
 ('Granollers', {'GF': 122, 'GC': 117, 'Puntos': 67}),
 ('Olot', {'GF': 132, 'GC': 144, 'Puntos': 64}),
 ('Sant Andreu', {'GF': 123, 'GC': 134, 'Puntos': 62}),
 ('Manlleu', {'GF': 106, 'GC': 118, 'Puntos': 57}),
 ('Prat', {'GF': 106, 'GC': 111, 'Puntos': 56}),
 ('Cerdanyola',

In [387]:
# PASO 9, Main, informe final

# El objetivo de la función es mostrar los resultados de la clasificación, ordenados y legibles, pudiendo responder a las preguntas que se nos han hecho.

def main_partidos():

    lineas_partidos = lectura()

    partidos = preparar_partidos(lineas_partidos)

    equipos = crear_equipos(partidos)

    equipos = procesar_goles(partidos)
    equipos = procesar_puntos(partidos, equipos)

    mas_goleador = equipo_mas_goleador(equipos)
    mas_goleado = equipo_mas_goleado(equipos)

    ranking_goles = ordenar_por_puntos(equipos)

    clasificacion = ordenar_clasificacion(equipos)

    print("\nEQUIPO MÁS GOLEADOR:")
    print("--------------------")
    print(mas_goleador[0], "->", mas_goleador[1], "goles")

    print("\nEQUIPO MÁS GOLEADO:")
    print("--------------------")
    print(mas_goleado[0], "->", mas_goleado[1], "goles recibidos")

    print("\nGOLES POR EQUIPO:")
    print("--------------------")
    for equipo, datos in ranking_goles:
     print(equipo, datos["GF"])

    print("\nCLASIFICACIÓN FINAL:")
    print("--------------------")

    for posicion, equipo in enumerate(clasificacion, start=1):
        print(
            posicion,
            equipo[0],
            "-",
            equipo[1]["Puntos"],
            "puntos"
        )

In [389]:
# COMPROBACIÓN DEL PASO 9

main_partidos()


EQUIPO MÁS GOLEADOR:
--------------------
Figueres -> 161 goles

EQUIPO MÁS GOLEADO:
--------------------
Vilafranca -> 172 goles recibidos

GOLES POR EQUIPO:
--------------------
Girona FC 139
Llagostera 159
Sabadell 141
Cornellà 147
RCD Espanyol 131
Figueres 161
Lleida Esportiu 129
Terrassa 147
FC Barcelona 125
Vilafranca 157
Badalona 134
Reus Deportiu 119
Nàstic de Tarragona 130
Granollers 122
Olot 132
Sant Andreu 123
Manlleu 106
Prat 106
Cerdanyola 117
Europa 93

CLASIFICACIÓN FINAL:
--------------------
1 Girona FC - 96 puntos
2 Llagostera - 94 puntos
3 Sabadell - 85 puntos
4 Cornellà - 82 puntos
5 RCD Espanyol - 80 puntos
6 Figueres - 79 puntos
7 Lleida Esportiu - 75 puntos
8 Terrassa - 74 puntos
9 FC Barcelona - 73 puntos
10 Vilafranca - 72 puntos
11 Badalona - 68 puntos
12 Reus Deportiu - 68 puntos
13 Nàstic de Tarragona - 68 puntos
14 Granollers - 67 puntos
15 Olot - 64 puntos
16 Sant Andreu - 62 puntos
17 Manlleu - 57 puntos
18 Prat - 56 puntos
19 Cerdanyola - 56 puntos
20 E